File: src/species-one.ts
import { Species } from './species.js';
import { Note } from './note.js';

export class SpeciesOne extends Species {
	protected noteOptions: number[] = [];
	protected previousIntervals: number[] = [];
	private lower: number[] = [];
	private upper: number[] = [];
	private count: number = 0;

	generateCounterpoint(cantusFirmus: Note[]): Note[] {
		return [];
	}

	chooseNextNote(): number {
		this.noteOptions = [];
		this.h_cannotCrossMelody();

		this.h_avoidDimFifth();
		this.h_noFourthOrSeventh();
		this.h_noSecondOrNinth();

		this.m_noParallelFifths();
		this.m_noSimilarFifths();
		this.m_noParallelOctaves();
		this.m_noSimilarOctaves();

		if (!(this.count % 4 === 0)) {
			this.h_removeEighth();
		}

		if (this.previousIntervals.length !== 0) {
			console.log("PreviousInterval: " + this.previousIntervals[this.previousIntervals.length - 1]);
		}

		const toChoose = Math.floor(Math.random() * this.noteOptions.length);
		const chosen = this.noteOptions[toChoose];

		return chosen;
	}

	writeImitativeTwoVoices(length: number = 8): void {
		this.lower = this.writeImitativeLowerVoice(length);
		for (let i = 0; i < length - 3; i++) {
			const temp = this.lower[i];
			this.upper.push(temp + 4);
		}
		this.upper.push(7);
		this.upper.push(8);
	}

	writeImitativeLowerVoice(length: number): number[] {
		const ImitativeLowerVoice: number[] = [];
		ImitativeLowerVoice.push(1);

		for (let i = 0; i < length - 3; i++) {
			let nextNote: number;
			const lastNote = ImitativeLowerVoice[ImitativeLowerVoice.length - 1];
			
			if (lastNote < -4) {
				nextNote = lastNote + this.pickImitativeUp() - 1;
			} else if (lastNote > 5) {
				nextNote = lastNote - this.pickImitativeDown() - 1;
			} else if (Math.random() < 0.5) {
				nextNote = lastNote - this.pickImitativeDown() - 1;
			} else {
				nextNote = lastNote + this.pickImitativeUp() - 1;
			}
			ImitativeLowerVoice.push(nextNote);
		}
		ImitativeLowerVoice.push(2);
		ImitativeLowerVoice.push(1);
		return ImitativeLowerVoice;
	}

	pickImitativeUp(): number {
		const rand = Math.floor(Math.random() * 9);
		switch (rand) {
			case 0:
				return 1;
			case 1:
			case 2:
			case 3:
			case 4:
				return 3;
			case 5:
			case 6:
				return 5;
			default:
				return 3;
		}
	}

	pickImitativeDown(): number {
		const rand = Math.floor(Math.random() * 7);
		switch (rand) {
			case 0:
				return 1;
			case 1:
			case 2:
			case 3:
			case 4:
				return 2;
			case 5:
			case 6:
				return 4;
			default:
				return 2;
		}
	}

	printImitativeCounterpoint(): void {
		console.log("Top:\t\t" + this.upper.join("\t"));
		console.log("Bottom:\t" + this.lower.join("\t"));
	}

	getImitativeUpper(): number[] {
		return this.upper;
	}

	getImitativeLower(): number[] {
		return this.lower;
	}

	protected h_cannotCrossMelody(): void {
		for (let i = this.noteBelow + 1; i < this.noteBelow + 9; i++) {
			this.noteOptions.push(i);
		}
	}

	protected h_avoidDimFifth(): void {
		if (this.noteBelow === 0 || this.noteBelow === 7) {
			const index = this.noteOptions.indexOf(this.noteBelow + 4);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
	}

	protected h_noFourthOrSeventh(): void {
		let index = this.noteOptions.indexOf(this.noteBelow + 3);
		if (index > -1) {
			this.noteOptions.splice(index, 1);
		}
		
		index = this.noteOptions.indexOf(this.noteBelow + 6);
		if (index > -1) {
			this.noteOptions.splice(index, 1);
		}
	}

	protected h_noSecondOrNinth(): void {
		let index = this.noteOptions.indexOf(this.noteBelow + 1);
		if (index > -1) {
			this.noteOptions.splice(index, 1);
		}
		
		index = this.noteOptions.indexOf(this.noteBelow + 8);
		if (index > -1) {
			this.noteOptions.splice(index, 1);
		}
	}

	protected h_removeEighth(): void {
		const index = this.noteOptions.indexOf(this.noteBelow + 7);
		if (index > -1) {
			this.noteOptions.splice(index, 1);
		}
	}

	protected m_noParallelFifths(): void {
		if ((this.noteBefore - 4) === this.noteBeforeAndBelow) {
			const index = this.noteOptions.indexOf(this.noteBelow + 4);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
	}

	protected m_noSimilarFifths(): void {
		if ((this.noteBeforeAndBelow > this.noteBelow) && ((this.noteBefore - 4) >= this.noteBeforeAndBelow)) {
			const index = this.noteOptions.indexOf(this.noteBelow + 4);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
		if ((this.noteBeforeAndBelow < this.noteBelow) && ((this.noteBefore - 4) <= this.noteBeforeAndBelow)) {
			const index = this.noteOptions.indexOf(this.noteBelow + 4);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
	}

	protected m_noParallelOctaves(): void {
		if ((this.noteBefore - 7) === this.noteBeforeAndBelow) {
			const index = this.noteOptions.indexOf(this.noteBelow + 7);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
	}

	protected m_noSimilarOctaves(): void {
		if ((this.noteBeforeAndBelow > this.noteBelow) && ((this.noteBefore - 7) >= this.noteBeforeAndBelow)) {
			const index = this.noteOptions.indexOf(this.noteBelow + 7);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
		if ((this.noteBeforeAndBelow < this.noteBelow) && ((this.noteBefore - 7) <= this.noteBeforeAndBelow)) {
			const index = this.noteOptions.indexOf(this.noteBelow + 7);
			if (index > -1) {
				this.noteOptions.splice(index, 1);
			}
		}
	}

	protected m_noSameNote(): void {
		const index = this.noteOptions.indexOf(this.noteBefore);
		if (index > -1) {
			this.noteOptions.splice(index, 1);
		}
	}

	protected m_onlyUse1Once(): void {
		this.previousIntervals.push(this.noteBefore - this.noteBeforeAndBelow + 1);

		const index = this.previousIntervals.indexOf(1);
		if (index > -1) {
			this.noteOptions.shift();
		}
	}
}

File: Music Project/SpeciesOne.cpp
#include "SpeciesOne.h"
#include "xorshift32.h"
#include <iostream>
#include <algorithm>


SpeciesOne::SpeciesOne()
{
}


SpeciesOne::~SpeciesOne()
{
}

int SpeciesOne::chooseNextNote() {
	noteOptions.clear();
	h_cannotCrossMelody(); // fills in a range above and equal to note below
	
	//For debugging
	//cout << "NoteOptions initialized:";
	//for (auto o : noteOptions) {
	//	cout << o << " ";
	//}
	//cout << endl;
	
	// Removes bad notes
	h_avoidDimFifth();
	h_noFourthOrSeventh();
	h_noSecondOrNinth();

	m_noParallelFifths();
	m_noSimilarFifths();
	m_noParallelOctaves();
	m_noSimilarOctaves();
	//m_onlyUse1Once(); // It is not working... so changed something else to get a similar effect. Code there now for reference for me later

	if (!(count % 4 == 0)) {
		h_removeEighth();
	}

	// For debugging
	//cout << "NoteOptions emptied:    ";
	//for (auto o : noteOptions) {
	//	cout << o << " ";
	//}
	//cout << endl;
	if (previousIntervals.size() != 0) {
		cout << "PreviousInterval: " << previousIntervals.at(previousIntervals.size() - 1) << endl;
	}

	int toChoose = Xorshift32::nextInt(noteOptions.size());
	int chosen = noteOptions.at(toChoose);
	
	//cout << "toChoose: " << toChoose << " NoteBelow: " << noteBelow << " nextNote: " << noteOptions.at(toChoose);
	return chosen;
}


		// Functions for imitative first species counterpoint

void SpeciesOne::writeImitativeTwoVoices(int length) {
	
	lower = writeImitativeLowerVoice(length);
	for (int i = 0; i < length - 3; i++) {
		int temp = lower.at(i);
		upper.push_back(temp + 4); // Imitative counter point a fifth above
	}
	upper.push_back(7);
	upper.push_back(8);
}

vector<int> SpeciesOne::writeImitativeLowerVoice(int length) {
	vector<int> ImitativeLowerVoice;
	ImitativeLowerVoice.push_back(1);

	for (int i = 0; i < length - 3; i++) {
		int nextNote;
		if (ImitativeLowerVoice.back() < -4) { // Sets bar for how low it will go
			nextNote = ImitativeLowerVoice.back() + pickImitativeUp() - 1;
		}
		else if (ImitativeLowerVoice.back() > 5) { // Sets bar for how high it will go
			nextNote = ImitativeLowerVoice.back() - pickImitativeDown() - 1;
		}
		else if (Xorshift32::nextFloat() < 0.5) {
			nextNote = ImitativeLowerVoice.back() - pickImitativeDown() - 1;
		}
		else {
			nextNote = ImitativeLowerVoice.back() + pickImitativeUp() - 1;
		}
		ImitativeLowerVoice.push_back(nextNote);
	}
	ImitativeLowerVoice.push_back(2);
	ImitativeLowerVoice.push_back(1);
	return ImitativeLowerVoice;
}

int SpeciesOne::pickImitativeUp() { // Returns the same note, a third, or a fifth above
	switch (Xorshift32::nextInt(9)) {
	case 0:
		return 1;
	case 1:
	case 2:
	case 3:
	case 4:
		return 3;
	case 5:
	case 6:
		return 5;
	default:
		return 3;
	}
}

int SpeciesOne::pickImitativeDown() { // Returns the same note, a second, or a fourth below
	switch (Xorshift32::nextInt(7)) {
	case 0:
		return 1;
	case 1:
	case 2:
	case 3:
	case 4:
		return 2;
	case 5:
	case 6:
		return 4;
	default:
		return 2;
	}
}

void SpeciesOne::printImitativeCounterpoint() {
	cout << "Top:" << "\t\t";
	for (auto note : upper) {
		cout << note << "\t";
	}
	cout << endl << "Bottom:" << "\t";
	for (auto note : lower) {
		cout << note << "\t";
	}
	cout << endl;
}

void SpeciesOne::h_cannotCrossMelody() {
	for (int i = noteBelow+1; i < noteBelow+9; i++) {
		noteOptions.push_back(i);
	}
}

void SpeciesOne::h_avoidDimFifth() {
	if ((noteBelow == 0) || (noteBelow == 7)) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 4);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
}

void SpeciesOne::h_noFourthOrSeventh() {
	// Find any perfect fourths
	vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 3);
	if (itr != noteOptions.end()) {
		noteOptions.erase(itr);	// Remove them
	}
	// Find any 7ths
	vector<int>::iterator itrr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 6);
	if (itrr != noteOptions.end()) {
		noteOptions.erase(itrr);	// Remove them
	}
}

void SpeciesOne::h_noSecondOrNinth() {
	// Find any 2nds
	vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 1);
	if (itr != noteOptions.end()) {
		noteOptions.erase(itr);	// Remove them
	}
	// Find any 9ths
	vector<int>::iterator itrr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 8);
	if (itrr != noteOptions.end()) {
		noteOptions.erase(itrr);	// Remove them
	}
}

void SpeciesOne::h_removeEighth() {
	vector<int>::iterator itrr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 7);
	if (itrr != noteOptions.end()) {
		noteOptions.erase(itrr);	// Remove them
	}
}

void SpeciesOne::m_noParallelFifths() {
	if ((noteBefore - 4) == noteBeforeAndBelow) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 4);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
}

void SpeciesOne::m_noSimilarFifths() {
	if ((noteBeforeAndBelow > noteBelow) && ((noteBefore - 4) >= noteBeforeAndBelow)) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 4);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
	if ((noteBeforeAndBelow < noteBelow) && ((noteBefore - 4) <= noteBeforeAndBelow)) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 4);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
}

void SpeciesOne::m_noParallelOctaves() {
	if ((noteBefore - 7) == noteBeforeAndBelow) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 7);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
}

void SpeciesOne::m_noSimilarOctaves() {
	if ((noteBeforeAndBelow > noteBelow) && ((noteBefore - 7) >= noteBeforeAndBelow)) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 7);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
	if ((noteBeforeAndBelow < noteBelow) && ((noteBefore - 7) <= noteBeforeAndBelow)) {
		// 5ths not allowed in this case
		vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBelow + 7);
		if (itr != noteOptions.end()) {
			noteOptions.erase(itr);
		}
	}
}

void SpeciesOne::m_noSameNote() {
	vector<int>::iterator itr = find(noteOptions.begin(), noteOptions.end(), noteBefore);
	if (itr != noteOptions.end()) {
		noteOptions.erase(itr);
	}
}

void SpeciesOne::m_onlyUse1Once() {
	previousIntervals.push_back(noteBefore - noteBeforeAndBelow + 1 );

	vector<int>::iterator itr = find(previousIntervals.begin(), previousIntervals.end(), 1);
	if (itr != previousIntervals.end()) {
		noteOptions.erase(noteOptions.begin());
	}
}

